src/marqo/core/index_management/vespa_application_package.py

```diff
@@ -0,0 +1,644 @@
+import xml.etree.ElementTree as ET
+
+from marqo.base_model import ImmutableBaseModel
+from marqo.core.exceptions import InternalError, OperationConflictError, IndexNotFoundError, IndexExistsError, \
+    ApplicationRollbackError
+from marqo.core.models import MarqoIndex
+import marqo.logging
+from marqo.vespa.exceptions import VespaError
+from marqo.vespa.vespa_client import VespaClient
+
+
+logger = marqo.logging.get_logger(__name__)
+
+
+class ServiceXml:
+    """
+    Represents a Vespa service XML file.
+    """
+    _CUSTOM_COMPONENT_BUNDLE_NAME = "marqo-custom-searchers"
+
+    def __init__(self, xml_str: str):
+        self._root = ET.fromstring(xml_str)
+        self._documents = self._ensure_only_one('content/documents')
+
+    def __repr__(self) -> str:
+        return f'ServiceXml({self.to_xml()})'
+
+    def to_xml(self) -> str:
+        # Add the namespace attribute to the root element
+        self._root.set('xmlns:deploy', 'vespa')
+        self._root.set('xmlns:preprocess', 'properties')
+
+        xml_declaration = '<?xml version="1.0" encoding="utf-8" ?>\n'
+        return xml_declaration + ET.tostring(self._root).decode('utf-8')
+
+    def _ensure_only_one(self, xml_path: str) -> ET.Element:
+        elements = self._root.findall(xml_path)
+
+        if len(elements) > 1:
+            raise InternalError(f'Multiple {xml_path} elements found in services.xml. Only one is allowed')
+        if len(elements) == 0:
+            raise InternalError(f'No {xml_path} element found in services.xml')
+
+        return elements[0]
+
+    def add_schema(self, name: str) -> None:
+        if self._documents.find(f'document[@type="{name}"]') is not None:
+            logger.warn(f'Schema {name} already exists in services.xml, nothing to add')
+        else:
+            new_document = ET.SubElement(self._documents, 'document')
+            new_document.set('type', name)
+            new_document.set('mode', 'index')
+
+    def remove_schema(self, name: str) -> None:
+        docs = self._documents.findall(f'document[@type="{name}"]')
+        if not docs:
+            logger.warn(f'Schema {name} does not exist in services.xml, nothing to remove')
+        else:
+            for doc in docs:
+                self._documents.remove(doc)
+
+    def config_components(self):
+        self._cleanup_container_config()
+        self._config_search()
+        self._config_index_setting_components()
+
+    def _cleanup_container_config(self):
+        """
+        Components config needs to be in sync with the components in the jar files. This method cleans up the
+        custom components config, so we can always start fresh. This assumes that the container section of the
+        services.xml file only has `node` config and empty `document-api` and `search` elements initially. Please
+        note that any manual config change in container section will be overwritten.
+        """
+        container_element = self._ensure_only_one('container')
```

Comment:
Our approach of fresh bootstrapping components config of container is based on following assumptions:
1. There's only one container element in the services.xml file. 
2. All configs in the container element except the nodes section is fully controlled by Marqo. 

These assumption might not be true in following cases:
- non-HA to HA migration in Marqo Cloud. A transiting state will have two container elements. In this case, we'd like to fail the bootstrapping process anyways to avoid entering any weird state. Also Marqo Cloud will make sure the upgrade of Marqo is blocked when processing migration
- OS customers could have multiple container clusters defined. (one use case is the separate the index and search traffic). Do we need to support this?
- OS customers could have other configuration in container section. should we keep those? 

If we have to support the cases above, I'll need to change the fresh bootstrapping process to an incremental one, which will make the logic more complicated. @farshidz , what's your thought?

Reply:
> All configs in the container element except the nodes section is fully controlled by Marqo.
This is fine but needs to be in Marqo docs, and also cloud migration guide

Reply:
will add it in the configuration.md file, and cloud migration guide